NV2A Improvements

JayFoxRox opened the pull request:
Ignore the branchname, it has grown a bit.
This includes a lot of changes. Currently online for review purposes.
The history was created after almost all of the code was written [I suck at "commit early, commit often"] so some commits might not be cherry-pickable.
## Feature list:
- More accurate texture modes [Not much to say about this one]
- Skinning [code skeleton for one of the 2 mixing modes, can't make it work properly and don't know how the second mode works]
- Bumpmap [Tested and working as far as I can tell]
- Alphakill [Untested]
- Texkill [Tested, works fine]
- Texgen [Missing 2 modes but otherwise seems to work fine]
- Texture matrices [Works]
- Texture addressing / wrap [Works but not all modes tested]
- Delta textures [Looks fine in the test code I ran]
- Better [565 instead 555 storage] RGB565 using ES2 compat [Untested, assumed to be working]
- Some smaller code changes

This requires the GL_ARB_ES2_compatibility extension for GL_RGB_565.

It also asserts on an unknown TEXTURE_FILTER_{ARGB}SIGNED state which might cause some games to not work anymore. I'm currently trying to figure out what those states do and if they are related to delta textures.
## TODO

Texgen and texture matrices only work using FFP at the moment but that will probably be solved in a second PR.

kthxbye

JayFoxRox committed:
More texture modes

JayFoxRox committed:
Alphakill, Texkill and Bumpmapping

JayFoxRox committed:
Add texture addressing

JayFoxRox committed:
New delta texture modes (req. GL_EXT_texture_snorm) and RGB_565
internal format (req. GL_ARB_ES2_compatibility)

JayFoxRox committed:
Better style in nv2a

JayFoxRox committed:
Texgen registers

JayFoxRox committed:
Normalization registers

JayFoxRox committed:
Bumpmapping Lum registers, better var name and fix for Bumpmapping

JayFoxRox committed:
Add TEXTURE_FILTER_?SIGNED assert, functionality not yet known (maybe delta tex?)

JayFoxRox committed:
Skinning registers

JayFoxRox committed:
Store Matrices (4x MV, 4x iMV, Proj)

JayFoxRox committed:
Texture matrix registers

JayFoxRox committed:
Prepare for vertex weights

JayFoxRox committed:
Cleanup PSH and alphakill

JayFoxRox committed:
Rename composite GLSL var to compositeMat

JayFoxRox committed:
Skinning, Texgen and Texture matrices WIP

espes commented on hw/xbox/nv2a.c:
```diff
@@ -3323,6 +3719,8 @@ static void pgraph_init(NV2AState *d)
     glextensions_init();
 
     assert(glo_check_extension("GL_EXT_texture_compression_s3tc"));
+    assert(glo_check_extension("GL_EXT_texture_snorm"));
```
SNORM textures are part of opengl 3.3 core

JayFoxRox replied:
cool! will remove this

espes commented on hw/xbox/nv2a.c:
```diff
@@ -3323,6 +3719,8 @@ static void pgraph_init(NV2AState *d)
 
     assert(glo_check_extension("GL_EXT_texture_compression_s3tc"));
+    assert(glo_check_extension("GL_EXT_texture_snorm"));
+    assert(glo_check_extension("GL_ARB_ES2_compatibility"));
```
Remind me, what's this for?

JayFoxRox replied:
GL_RGB_565 because we used to store 565 in 555 which might not be enough for some applications.

JayFoxRox committed:
Remove req. for GL_EXT_texture_snorm (Part of GL3.3)

JayFoxRox committed:
Explanation for GL extensions

JayFoxRox committed:
Add some FIXME

JayFoxRox committed:
Fix some parts of skinning

JayFoxRox committed:
Style + FIXME for assert

JayFoxRox committed:
Fix skinning normals and cleanup

JayFoxRox commented:
~~Can be merged now unless there are any complains :)~~
Style should be okay, some lines over 80 chars but pretty much all of it is shader code which should not be broken for grep-reasons according to style doc.

espes commented on hw/xbox/nv2a.c:
```diff
@@ -1972,6 +2091,30 @@ static uint8_t* convert_texture_data(const TextureShape s,
+                 * signed 565 to signed 888 (hopefully)
+                 */
+                v = (rgb565 & 0xF800) >> 8;
+                pixel[0] = (v & 0x80) ? *(int8_t*)&v * 16 / 0x80
```
I don't think casting v through a pointer is necessary.

Why is this doing 565 -> 888 instead of 655 -> 888?

Also the usual sign extension strategy is (int8_t)(x << (8-k)) >> (8-k)

Should this format even be signed?

JayFoxRox replied:
A simple oversight! I never noticed it's 655! Will fix that later.
(It's a delta texture so I can't use apitrace to debug it and the Bumpmapping looked okay)
~~I'll also use your sign extension - I struggled with that but was to emberassed to ask~~
//Edit: This is not a simple sign extension but remapping 5 [-16,15] / 6 [-32,31] bit signed to 8 bit signed [-128,127].

espes commented on hw/xbox/nv2a_vsh.h:
```diff
@@ -24,6 +24,25 @@
 
 #include "qapi/qmp/qstring.h"
 
+enum Texgen {
```
nit: make this VshTexgen, VshSkinning?

JayFoxRox replied:
Yes.
I just called it that because there was also "enum AlphaFunc".
That should probably be renamed to PshAlphaFunc then, too?

Thanks for the feedback!

espes replied:
Probably, yeah. Don't worry too much about it, we'll make a pass refactoring the vsh and psh stuff eventually.

Thanks!

JayFoxRox committed:
Give VSH and PSH types a prefix

JayFoxRox committed:
Fix delta-R6G5B5 code (incl. new mapping algorithm)

JayFoxRox commented:
I rewrote the R6G5B5 mapping algorithm (A couple of times as you can see ;) ), assuming the 6 bit value is in the uint16_t MSB.

It seems these textures are also called L6V5U5 and U8V8. A Description can be found here: https://msdn.microsoft.com/en-us/library/windows/desktop/bb172558%28v=vs.85%29.aspx#Mixed_Formats.
L controls the luminance and should probably be unsigned, however, U and V are part of the signed format list and they offset the bumpmap shadows/highlights in each direction. Wether the signedness of each channel is register controlled is unknown to me. I added asserts to the TEX_FILTER_{ARGB}SIGNED but those seem unrelated.
I'm not 100% sure it's always signed yet. I noted that by putting a FIXME comment near the texture descriptor.

The new algorithm works like this:

The 6 bit luminance is mapped from [0,63] -> [0,127] so it will return as [0.0,1.0] in the shader.
For the 5 bit channels U (shader r) and V (shader g) the following is done:
First the sign bit is flipped to map values: [-16,15] -> [0,31]
Now we can multiply by 255 and divide by 31: [0,31] -> [0,7905] -> [0,255]
And finally map to the new range by subtracting 128: [0,255] -> [-128,127]
That will equal [-1.0,1.0] in the shader.

I also renamed those PSH and VSH types.

## I believe this can REALLY be merged now. Thanks for the constructive feedback!

JayFoxRox committed:
Fix Texgen by doing skinning fixup later

JayFoxRox commented:
Tiny change to fix texgen. Merge please.

espes commented:
<3
